### DISCUSSION TITLE
Plotting data missing point

### DISCUSSION THREAD

**Discussion by gallacci97** [nasa/openmct#4318] [Community Q&A] [answered]
Hi everyone, 
I'm developing an application in order to visualize data coming from a satellite.
I noticed that data are not plotted as they arrive to openmct and I can't find the reasons. A momentary solution is to force quering to my historical server for data.
As you can see in video data are retrieved correctly when I change to Table view.

This is the plugin I'm using:

` function RealtimeTelemetryMQTTPluginArgo() {
    return function (openmct) {
        var client  = mqtt.connect('ws://mqtt_server', {
            host: "ws://mqtt_server", 
            protocol: "mqtt",
            port:8081,
            clientId: "clientid",
            username: 'username',
            password: 'password'
            
        });

        var listener = {};
        var history = {};
        var h_promises = {};
        client.on('connect', function () {
            console.log('Client connesso');
        });

        client.on('error', function(error){
            console.log(error);
        });
        
        client.on('message', function (topic, payload) {
            // message is Buffer
            console.log([topic, payload].join(": "));
            point = JSON.parse(payload);
            
            if ("history" in point){
                console.log("history ricevuta");
                if (point.history in history){
                    delete history[point.history];

                }
                history[point.history] = point.Value;
                h_promises[point.history] = true;
            }else{
                if(listener[point.id]){
                    listener[point.id](point);
                }
            }
            //console.log(point.toString())
            

        });

        var provider = {
            supportsSubscribe: function (domainObject) {
                return domainObject.type === 'spacecraft.telemetry';
            },
            subscribe: function (domainObject, callback) {
                listener[domainObject.identifier.key] = callback;
                var topic = domainObject.identifier.key.replace(/\./g,"/");
                client.subscribe(topic, qos=1, function (err,granted) {
                    if (!err) {
                      //client.publish('test/topic', 'Hello mqtt')
                      console.log("Client iscritto");
                      console.log(topic);
                    }
                    console.log(granted.toString());
                    
                })

                return function unsubscribe() {
                    delete listener[domainObject.identifier.key];
                    var topic = domainObject.identifier.key.replace(/\./g,"/");
                    client.unsubscribe(topic);
                };
            },

            supportsRequest: function (domainObject){
                return domainObject.type == 'spacecraft.telemetry';
            },
            request: function (domainObject, options) {
                var t_start = options.start;
                var t_end = options.end;
                var key = domainObject.identifier.key;
                const message = {TYPE: 1, id: key, start: t_start, end: t_end}; 
                var index = key.indexOf(".");
                var topiclisten = key.slice(0, index) + ".history.clientid" + key.slice(index);
                topiclisten = topiclisten.replace(/\./g,"/");
                client.subscribe(topiclisten, qos=1, function (err,granted) {
                    if (!err) {
                      //client.publish('test/topic', 'Hello mqtt')
                      console.log("Client iscritto per history");
                      console.log(topiclisten);
                    }
                    console.log(granted.toString());
                    
                })
                var topicsend = key.slice(0, index) + ".history.clientid.clientid"
                topicsend = topicsend.replace(/\./g,"/");
                h_promises[key] = false;
                console.log(JSON.stringify(message));
                client.publish(topicsend, JSON.stringify(message));

                const  getHistory = (k) => {
                    return new Promise ((resolve, reject) => {
                        (function waitForHistory(){
                            if(h_promises[k] == true){
                                if(h_promises[k].length == 0 ) reject("no data")
                                return resolve(history[key]);
                            } 
                            setTimeout(waitForHistory, 50);
                        })();
                    })
                };
                

                /*getHistory(key).then( resp => {
                    console.log(resp);
                    client.unsubscribe(topiclisten);
                    return resp;
                }).catch(function (err) {
                    console.log(err);
                  });*/
                return getHistory(key);
            }
        };
        
        openmct.telemetry.addProvider(provider);
    }
}
`

Do you have idea of how to solve this problem or did you encountered in the past a similar behavior?

Thank you

https://user-images.githubusercontent.com/58908937/137290252-be99fcb8-fc6d-4782-ac34-29ca1fb360f5.mp4

**Comment by akhenry** [comment, marked as answer]
It looks to me as though the real-time data has timestamps that are in the future (according to Open MCT). This explains why the plot is showing a line that appears to go off the screen to the right, even though there are no data points plotted (it is actually plotting the data points, they're just off-screen).

Here's an example from the video you provided (thanks! This was really helpful in diagnosing the issue):
![Screen Shot 2021-10-22 at 9 16 49 AM](https://user-images.githubusercontent.com/4215777/138489395-98b17185-4fee-4735-bf0f-604f719268fb.png)

Note that the last timestamp in the data is later than latest time being considered by the time conductor. The Open MCT clock is based on the local time on the machine that the browser is running on. This could be caused by clock drift in either the client, or the machine that is producing the telemetry timestamps (eg. if there is a VM involved), or it could be due to leap-second handling. JavaScript time does not include leap-seconds, so this needs to be corrected for.

**Reply by gallacci97** [reply]
Thank you really much, as the code worked fine for a while I did not think about clock drift. I introduced synchronization control messages in order to automatically take in account time disagreements
